ecs-init/cache/cache.go

```diff
@@ -36,28 +39,41 @@ const (
 
 // Downloader is responsible for cache operations relating to downloading the agent
 type Downloader struct {
-	getter   httpGetter
-	fs       fileSystem
-	metadata instanceMetadata
-	region   string
+	s3downloader s3Downloader
+	fs           fileSystem
+	metadata     instanceMetadata
+	region       string
 }
 
 // NewDownloader returns a Downloader with default dependencies
-func NewDownloader() *Downloader {
+func NewDownloader() (*Downloader, error) {
 	downloader := &Downloader{
-		getter: customGetter,
-		fs:     &standardFS{},
+		fs: &standardFS{},
 	}
 
 	// If metadata cannot be initialized the region string is populated with the default value to prevent future
 	// calls to retrieve the region from metadata
 	sessionInstance, err := session.NewSession()
 	if err != nil {
+		log.Debugf("Get error when initializing session instance: %v. Use default region: %s", 
+			err, config.DefaultRegionName)
 		downloader.region = config.DefaultRegionName
-		return downloader
+	} else {
+		// metadata is only used for retrieving the user's region
+		downloader.metadata = ec2metadata.New(sessionInstance)
 	}
-	downloader.metadata = ec2metadata.New(sessionInstance)
-	return downloader
+
+	log.Debugf("Setting region for s3 client to: %s", downloader.getBucketRegion())
+	s3sessionInstance, err := session.NewSession(&aws.Config{
```

Comment:
Why do we need 2 invocations of the `session.NewSession()` call? Seems like we should be fine with creating a new session just once?

Reply:
I think we actually need two session instances because the session used for downloading from s3 needs a region to initialize, and that region is retrieved from the metadata, which needs a session instance... I don't find a method to set a region for session/download client

Reply:
I see. A minor nit here: can you rename `s3sessionInstance` to `s3Session`?

Reply:
ok